### PR TITLE
Refactor chunking algorithm changes

### PR DESCRIPTION
**Pull request by rgmz**
### Description:
This is a follow-up to #1665. 

Changes include:
- Moving the logic into a separate function and adding test cases
- Also implementing the changes in `reader.go`.
- Using `bufio.Reader` in `directory.go` to speed up execution

### Checklist:

* [x] Does your PR pass tests?
* [x] Have you written new tests for your changes?
* [x] Have you lint your code locally prior to submission?

### CODE REVIEW

File: detect/directory.go [hunk 1]
```diff
@@ -49,64 +50,33 @@ func (d *Detector) DetectFiles(paths <-chan sources.ScanTarget) ([]report.Findin
 				}
 			}
 
-			// Buffer to hold file chunks
-			buf := make([]byte, chunkSize)
-			totalLines := 0
+			var (
+				// Buffer to hold file chunks
+				reader     = bufio.NewReader(f)
```
Comment (rgmz) [inline detect/directory.go:55]: Using #1432 as a test case. @ahrav @zricethezav any obvious errors with my use of `bufio.Reader`?

**Before**
```
12:33PM INF scanned ~104857630 bytes (104.86 MB) in 5.63s
12:33PM WRN leaks found: 1
```
**After**
```
1:05PM INF scanned ~104857630 bytes (104.86 MB) in 1.42s
1:05PM WRN leaks found: 1
```

File: detect/directory.go [hunk 2]
```diff
@@ -49,64 +50,32 @@ func (d *Detector) DetectFiles(paths <-chan sources.ScanTarget) ([]report.Findin
...
 
-						// Read additional data into a temporary buffer
-						m, readErr := f.Read(tempBuf)
-						if m > 0 {
-							peekBuf.Write(tempBuf[:m])
+				// "Callers should always process the n > 0 bytes returned before considering the error err."
+				// https://pkg.go.dev/io#Reader
+				if n > 0 {
+					// Only check the filetype at the start of file.
+					if totalLines == 0 {
```
Comment (rgmz) [inline detect/directory.go:66]: This doesn't need to be run after the first chunk.

File: detect/directory.go [hunk 3]
```diff
@@ -145,3 +114,49 @@ func (d *Detector) DetectFiles(paths <-chan sources.ScanTarget) ([]report.Findin
...
+		if lastChar == '\n' {
+			newlineCount++
+
+			// Stop if two consecutive newlines are found
+			if newlineCount >= 2 {
+				break
+			}
+		} else if lastChar == '\r' || lastChar == ' ' || lastChar == '\t' {
+			// The presence of other whitespace characters (`\r`, ` `, `\t`) shouldn't reset the count.
+			// (Intentionally do nothing.)
```
Comment (rgmz) [inline detect/directory.go:141]: This handles https://github.com/gitleaks/gitleaks/pull/1665#issuecomment-2539410608

File: detect/directory.go (hunk 1 above)
Comment (ahrav) [reply detect/directory.go:55]: This looks correct to me. I agree—it does feel a bit odd to use all three functions, but if I’m understanding this correctly, they each serve different purposes, so it actually makes sense.

Comment (rgmz) [reply detect/directory.go:55]: I'm concerned that `f.Read` and `r.ReadByte` may not place nicely. E.g., after the first iteration does calling `f.Read` also progress the internal reader used by bufio?

The number of bytes read (from #1670) appears to be consistent. 🤷

**Pull request merged**